File: day16/day16.py
d = set()
with open('input.txt') as f:
    l = list('abcdefghijklmnop')
    # l = list('abcde')

    
    # print(''.join(l))
    for i in range(16):
        # if i % 1000 == 0:
            # print(i)
        f.seek(0)
        for instr in f.read().split(','):
            if instr[0] == 's':
                num = int(instr[1:])
                l = l[-num:] + l[:-num]
            if instr[0] == 'x':
                a, b = instr[1:].split('/')
                l[int(a)], l[int(b)] = l[int(b)], l[int(a)]
            if instr[0] == 'p':
                p1, p2 = instr[1], instr[3]
                p1i, p2i = l.index(p1), l.index(p2)
                l[p1i], l[p2i] = p2, p1

        if ''.join(l) in d:
            print(i)
        d.add(''.join(l))

    print(''.join(l))

File: day16/day16.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

int main() {
  string l = "abcdefghijklmnop";
  for (int i = 0; i < 1000000000; ++i) {
    if (i % 1000 == 0) {
      cout << i << endl;
    }
    ifstream f;
    f.open("input.txt");
    char instr;
    while (true) {
      cin >> instr;
      if (instr == 's') {
        int num;
        cin >> num;

      }
      else if (instr == 'x') {
        int a, b;
        char slash;
        cin >> a >> slash >> b;
        swap(l[a], l[b]);
      }
      else if (instr == 'p') {
        char a, b;
        char slash;
        cin >> a >> slash >> b;
        swap(l[l.find(a)], l[l.find(b)]);
      }
      else {
        break;
      }
    }
  }
}

